### ISSUE TITLE
External Authentication can create new headers, but not overwrite existing headers

### ISSUE THREAD

**Issue by joatmon** [datawire/ambassador#476]
**Describe the bug**
Allowed headers from the external authentication response are copied into the upstream request if and only if that header does not already exist.

**To Reproduce**
Sample configuration:
```
apiVersion: ambassador/v0
kind:  AuthService
name:  authentication
auth_service: "hostname:8001"
path_prefix: "/authenticate"
allowed_headers:
- "Set-cookie"
- "Authorization"
```
I sent in a request that had the value 'jk' in the Authorization header. Here are the envoy logs for the response from the external authentication and the resulting upstream request with the original Authorization header.
```
[2018-05-30 18:30:09.281][21][debug][router] source/common/router/router.cc:573] [C0][S16497426937251309152] upstream headers complete: end_stream=false
[2018-05-30 18:30:09.281][21][debug][http] source/common/http/async_client_impl.cc:85] async http request response headers (end_stream=false):
[2018-05-30 18:30:09.281][21][debug][http] source/common/http/async_client_impl.cc:88]   'date':'Wed, 30 May 2018 18:30:09 GMT'
[2018-05-30 18:30:09.281][21][debug][http] source/common/http/async_client_impl.cc:88]   'access-control-allow-origin':'*'
[2018-05-30 18:30:09.282][21][debug][http] source/common/http/async_client_impl.cc:88]   'content-encoding':'gzip'
[2018-05-30 18:30:09.282][21][debug][http] source/common/http/async_client_impl.cc:88]   'cache-control':'no-store, no-cache, must-revalidate, proxy-revalidate'
[2018-05-30 18:30:09.282][21][debug][http] source/common/http/async_client_impl.cc:88]   'pragma':'no-cache'
[2018-05-30 18:30:09.282][21][debug][http] source/common/http/async_client_impl.cc:88]   'expires':'0'
[2018-05-30 18:30:09.282][21][debug][http] source/common/http/async_client_impl.cc:88]   'content-type':'application/json'
[2018-05-30 18:30:09.282][21][debug][http] source/common/http/async_client_impl.cc:88]   'authorization':'Bearer eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJ2dVdPcGJkQW5ZSzl0OGx4cXRYSHBMOGVLeDlJQzhDRnc1RUJVdXpXNm40In0.eyJqdGkiOiJiM2NlYzE1Zi1hNmYxLTQ1MTgtOTNkOS05ODY5ZmI1ZDcxMDIiLCJleHAiOjE1Mjc3MDg1OTIsIm5iZiI6MCwiaWF0IjoxNTI3NzA0OTkyLCJpc3MiOiJodHRwOi8va2V5Y2xvYWs6ODA4MC9hdXRoL3JlYWxtcy9tYXN0ZXIiLCJhdWQiOiJ1c2VyLXNlcnZpY2UiLCJzdWIiOiIxY2I0NjFhNy0xMWE2LTQwZWMtYmRiYi1iZGNmNTdjNDQ1NDMiLCJ0eXAiOiJCZWFyZXIiLCJhenAiOiJ1c2VyLXNlcnZpY2UiLCJub25jZSI6ImE1cURzN0YxRUoyd3lxMmlxckhKc0J2bW44VzZhT1lkcUpwcFl4TXVWZTQiLCJhdXRoX3RpbWUiOjE1Mjc3MDQ5MjQsInNlc3Npb25fc3RhdGUiOiI2YTU5ZTg3YS0zMjBiLTQ1Y2EtYWQyYS0zMGQ1OTZiNDgwMzQiLCJhY3IiOiIwIiwiYWxsb3dlZC1vcmlnaW5zIjpbXSwicmVhbG1fYWNjZXNzIjp7InJvbGVzIjpbImNyZWF0ZS1yZWFsbSIsImFkbWluIiwidW1hX2F1dGhvcml6YXRpb24iXX0sInJlc291cmNlX2FjY2VzcyI6eyJtYXN0ZXItcmVhbG0iOnsicm9sZXMiOlsidmlldy1pZGVudGl0eS1wcm92aWRlcnMiLCJ2aWV3LXJlYWxtIiwibWFuYWdlLWlkZW50aXR5LXByb3ZpZGVycyIsImltcGVyc29uYXRpb24iLCJjcmVhdGUtY2xpZW50IiwibWFuYWdlLXVzZXJzIiwicXVlcnktcmVhbG1zIiwidmlldy1hdXRob3JpemF0aW9uIiwicXVlcnktY2xpZW50cyIsInF1ZXJ5LXVzZXJzIiwibWFuYWdlLWV2ZW50cyIsIm1hbmFnZS1yZWFsbSIsInZpZXctZXZlbnRzIiwidmlldy11c2VycyIsInZpZXctY2xpZW50cyIsIm1hbmFnZS1hdXRob3JpemF0aW9uIiwibWFuYWdlLWNsaWVudHMiLCJxdWVyeS1ncm91cHMiXX0sImFjY291bnQiOnsicm9sZXMiOlsibWFuYWdlLWFjY291bnQiLCJtYW5hZ2UtYWNjb3VudC1saW5rcyIsInZpZXctcHJvZmlsZSJdfX0sInByZWZlcnJlZF91c2VybmFtZSI6ImFkbWluIn0.YjFEkTyFMKdN4xHhSOUP-2E2wsD5GTE4Dz5XsROkTX89K2HHUacM6-V6LxKhxOsqnk40mh2y0rjYdDcTrsY-F_ZY_uEOPitIuiLnuwzMGZkYAxaNatw-Vw0QR3jTzdOJhG9NkLLdjZbunko-TXAjh6Klmfn3qrzHYJODCdcFHcyJcUf1dN8SQti_XEUadxAEGGU5lmDL7ihmrJVAOMpd-FdegNe2f4MO-17yzrbwhwTswAmH27kSzlLibxE6YaScU7E35Z8FBhqEbCTr_oX9Z0MheO1XAxxg2D8N_yCu52BnlzLQj0uFV7Ct0O30EBciJI00edHcYvWBp5G_4wKrZQ'
[2018-05-30 18:30:09.282][21][debug][http] source/common/http/async_client_impl.cc:88]   'transfer-encoding':'chunked'
[2018-05-30 18:30:09.282][21][debug][http] source/common/http/async_client_impl.cc:88]   'server':'Jetty(9.4.z-SNAPSHOT)'
[2018-05-30 18:30:09.282][21][debug][http] source/common/http/async_client_impl.cc:88]   ':status':'200'
[2018-05-30 18:30:09.282][21][debug][http] source/common/http/async_client_impl.cc:88]   'x-envoy-upstream-service-time':'3'
[2018-05-30 18:30:09.283][21][debug][client] source/common/http/codec_client.cc:81] [C107] response complete
[2018-05-30 18:30:09.283][21][debug][filter] source/common/http/filter/extauth.cc:223] [C103][S5397472963650963421] ExtAuth accepting request
[2018-05-30 18:30:09.283][21][debug][router] source/common/router/router.cc:248] [C103][S5397472963650963421] cluster 'cluster_system' match for URL '/api/system/systems'
[2018-05-30 18:30:09.284][21][debug][router] source/common/router/router.cc:293] [C103][S5397472963650963421]   ':authority':'localhost:9999'
[2018-05-30 18:30:09.284][21][debug][router] source/common/router/router.cc:293] [C103][S5397472963650963421]   'accept':'application/json'
[2018-05-30 18:30:09.284][21][debug][router] source/common/router/router.cc:293] [C103][S5397472963650963421]   'authorization':'jk'
```
If the incoming request does not have an `Authorization` header at all, then envoy copies the Authorization header provided by the External Authorization service as expected.

**Expected behavior**
Allowed Headers returned by the External Authorization service should replace existing headers in the upstream request.

**Versions (please complete the following information):**
 - Ambassador: 0.33.1

**Additional context**
Here is an example use case: an incoming request has an expired JWT in the `Authorization` header. My external authorization service has the ability to refresh that token but cannot do so because it cannot provide new content for a header that already exists in the request.

**Comment by kflynn** [comment]
Confirmed, sadly enough -- if you're using some arbitrary `X-` header it works fine, but if you're trying to overwrite `Authorization`, it doesn't. This is almost certainly a bug in the Envoy `extauth` filter itself related to Envoy's fastpath vs slowpath header handling. As such it'll require a filter change and a new `ambassador-envoy` build (neither of which require an Envoy release, FTR).

**Comment by richarddli** [comment]
The problem is identified, and fixed it in https://github.com/datawire/envoy/commit/d1579ff3bf9a3cd854c7d28ba7c39c4aa3f83b89 and https://github.com/datawire/envoy/commit/6a44698122d7c992beab89915d5caf7b6f911fe8. We need do a full regression test and ship the fix.

**Comment by richarddli** [comment]
Fixed in 0.34.2.
